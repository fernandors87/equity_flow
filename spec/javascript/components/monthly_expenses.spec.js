import MonthlyExpenses, { Body, Footer, Header, Row } from 'components/monthly_expenses'
import Account from 'models/account'
import { List } from 'immutable'
import Split from 'models/split'
import moment from 'moment'

describe('MonthlyExpenses', () => {

  const accounts = List.of(
    Account({ id: 2, name: 'a2', type: 'expense' }),
    Account({ id: 1, name: 'a1', type: 'expense' }),
    Account({ id: 3, name: 'a3', type: 'income' })
  )

  const months = List.of(moment.utc('2017-01-01'), moment.utc('2017-03-01'))

  const splits = List.of(
    Split({ id: 1, date: moment.utc('2017-01-01'), value: 11.6 }),
    Split({ id: 2, date: moment.utc('2017-01-31'), value: 12.6 }),
    Split({ id: 3, date: moment.utc('2017-02-15'), value: 13.6 }),
    Split({ id: 4, date: moment.utc('2017-03-01'), value: 14.6 }),
    Split({ id: 5, date: moment.utc('2017-03-30'), value: 15.6 })
  )

  describe('Header', () => {

    it('render the fragment', () => {
      const wrapper = shallow(<Header months={ months } />)
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('Body', () => {

    it('render the fragment', () => {
      const wrapper = shallow(<Body accounts={ accounts } months={ months } splits={ splits } />)
      expect(wrapper).toMatchSnapshot()
    })

    it('render rows for each account', () => {
      const wrapper = shallow(<Body accounts={ accounts } />)
      const actual = wrapper.find('Row')
      expect(actual.length).toEqual(3)
    })

    it('sort the accounts in alphabetical order', () => {
      const wrapper = shallow(<Body accounts={ accounts } />)
      const actual = wrapper.find('Row').map(x => x.props().label)
      expect(actual).toEqual(['a1', 'a2', 'a3'])
    })
  })

  describe('Row', () => {

    it('render the fragment', () => {
      const wrapper = shallow(<Row label="Blah" months={ months } splits={ splits } />)
      expect(wrapper).toMatchSnapshot()
    })

    it('group splits into the given months', () => {
      const wrapper = shallow(<Row label="Blah" months={ months } splits={ splits } />)
      const actual = wrapper.find('.value').map(x => x.props().children)
      expect(actual).toEqual(['24', '30'])
    })

    it('calculate totals', () => {
      const wrapper = shallow(<Row label="Blah" months={ months } splits={ splits } />)
      const total = wrapper.find('.total').text()
      const avg = wrapper.find('.avg').text()
      expect(total).toEqual('68')
      expect(avg).toEqual('34')
    })
  })

  describe('Footer', () => {

    it('render the fragment', () => {
      const wrapper = shallow(<Footer months={ months } splits={ splits } />)
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('MonthlyExpenses', () => {

    it('render the component', () => {
      const startDate = moment.utc('2018-03-01')
      const endDate = moment.utc('2018-04-15')
      const wrapper = shallow(<MonthlyExpenses accounts={ accounts } splits={ splits } startDate={ startDate } endDate={ endDate } />)
      expect(wrapper).toMatchSnapshot()
    })
  })
})
