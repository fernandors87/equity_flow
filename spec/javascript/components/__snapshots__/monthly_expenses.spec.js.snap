// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthlyExpenses Body render the fragment 1`] = `
<tbody>
  <Row
    key="1"
    label="a1"
    months={
      Immutable.List [
        "2017-01-01T00:00:00.000Z",
        "2017-03-01T00:00:00.000Z",
      ]
    }
    splits={Immutable.List []}
  />
  <Row
    key="2"
    label="a2"
    months={
      Immutable.List [
        "2017-01-01T00:00:00.000Z",
        "2017-03-01T00:00:00.000Z",
      ]
    }
    splits={Immutable.List []}
  />
  <Row
    key="3"
    label="a3"
    months={
      Immutable.List [
        "2017-01-01T00:00:00.000Z",
        "2017-03-01T00:00:00.000Z",
      ]
    }
    splits={Immutable.List []}
  />
</tbody>
`;

exports[`MonthlyExpenses Footer render the fragment 1`] = `
<tfoot>
  <Row
    label="Total"
    months={
      Immutable.List [
        "2017-01-01T00:00:00.000Z",
        "2017-03-01T00:00:00.000Z",
      ]
    }
    splits={
      Immutable.List [
        Immutable.Split {
          "id": 1,
          "account_id": null,
          "deal_id": null,
          "value": 11.6,
          "date": "2017-01-01T00:00:00.000Z",
        },
        Immutable.Split {
          "id": 2,
          "account_id": null,
          "deal_id": null,
          "value": 12.6,
          "date": "2017-01-31T00:00:00.000Z",
        },
        Immutable.Split {
          "id": 3,
          "account_id": null,
          "deal_id": null,
          "value": 13.6,
          "date": "2017-02-15T00:00:00.000Z",
        },
        Immutable.Split {
          "id": 4,
          "account_id": null,
          "deal_id": null,
          "value": 14.6,
          "date": "2017-03-01T00:00:00.000Z",
        },
        Immutable.Split {
          "id": 5,
          "account_id": null,
          "deal_id": null,
          "value": 15.6,
          "date": "2017-03-30T00:00:00.000Z",
        },
      ]
    }
  />
</tfoot>
`;

exports[`MonthlyExpenses Header render the fragment 1`] = `
<thead>
  <tr>
    <th>
      Account
    </th>
    <th
      key="2017-01-01T00:00:00Z"
    >
      2017-01
    </th>
    <th
      key="2017-03-01T00:00:00Z"
    >
      2017-03
    </th>
    <th>
      Sum
    </th>
    <th>
      Avg
    </th>
  </tr>
</thead>
`;

exports[`MonthlyExpenses MonthlyExpenses render the component 1`] = `
<table
  className="monthly-expenses table table-hover"
>
  <Header
    months={
      Immutable.List [
        "2018-03-01T00:00:00.000Z",
        "2018-04-01T00:00:00.000Z",
      ]
    }
  />
  <Body
    accounts={
      Immutable.List [
        Immutable.Account {
          "id": 2,
          "name": "a2",
          "description": null,
          "type": "expense",
          "parent_id": null,
          "level": null,
          "full_name": null,
          "children": Immutable.List [],
        },
        Immutable.Account {
          "id": 1,
          "name": "a1",
          "description": null,
          "type": "expense",
          "parent_id": null,
          "level": null,
          "full_name": null,
          "children": Immutable.List [],
        },
      ]
    }
    months={
      Immutable.List [
        "2018-03-01T00:00:00.000Z",
        "2018-04-01T00:00:00.000Z",
      ]
    }
    splits={Immutable.List []}
  />
  <Footer
    months={
      Immutable.List [
        "2018-03-01T00:00:00.000Z",
        "2018-04-01T00:00:00.000Z",
      ]
    }
    splits={Immutable.List []}
  />
</table>
`;

exports[`MonthlyExpenses Row render the fragment 1`] = `
<tr>
  <td
    className="category"
  >
    Blah
  </td>
  <td
    className="value"
    key="2017-01-01T00:00:00Z"
  >
    24
  </td>
  <td
    className="value"
    key="2017-03-01T00:00:00Z"
  >
    30
  </td>
  <td
    className="summary total"
  >
    68
  </td>
  <td
    className="summary avg"
  >
    34
  </td>
</tr>
`;
